tatanka/client/orderbook/orderbook.go

```diff
@@ -0,0 +1,169 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package orderbook
+
+import (
+	"errors"
+	"sort"
+	"sync"
+
+	"decred.org/dcrdex/tatanka/tanka"
+)
+
+// OrderBooker specifies the orderbook interface.
+type OrderBooker interface {
+	OrderIDs() [][32]byte
+	Order(id [32]byte) *tanka.OrderUpdate
+	Orders(ids [][32]byte) []*tanka.OrderUpdate
+	FindOrders(filter *OrderFilter) []*tanka.OrderUpdate
+	AddUpdate(*tanka.OrderUpdate) error
+	Delete(id [32]byte)
+}
+
+// OrderFilter is used when searching for orders.
+type OrderFilter struct {
+	IsSell *bool
+	Check  func(*tanka.OrderUpdate) (ok, done bool)
+}
+
+var _ OrderBooker = (*OrderBook)(nil)
+
+// OrderBook holds the orderbook.
+type OrderBook struct {
+	mtx  sync.RWMutex
+	book map[[32]byte]*tanka.OrderUpdate
+	// buys and sells use the same pointers as the book and are sorted
+	// with the best trade first.
+	buys, sells []*tanka.OrderUpdate
+}
+
+// NewOrderBook created a new orderbook.
+func NewOrderBook() *OrderBook {
+	return &OrderBook{
+		book: make(map[[32]byte]*tanka.OrderUpdate),
+	}
+}
+
+// OrderIDs returns all order ids.
+func (ob *OrderBook) OrderIDs() [][32]byte {
```

Comment:
What's the intent for this method?

Reply:
I imagine for communications with other nodes, you would tell them what orders you have in your db, then they would ask for the full info on ones they do not have. But maybe it would just be the best X order ids?

Reply:
It does raise another question about how those messages to the new sub would be sent. They are p2p, but they don't really need e2e encryption (although they could). Do they still go across the subject channel, as a special broadcast type with one recipient?

Or perhaps standing order owners just broadcast an order update with a fresh stamp every "epoch" if they see there are new users?